CONTRIBUTING.md

```diff
@@ -1,4 +1,68 @@
-
+# The Build
+
+## Values
+
+We put a huge premium on a *very, very fast, perfectly consistent* build. 
+  * A typical application code change to a "leaf" module should take no more
+    than ~30 seconds to result a build status message that includes a link to a deployed version of the code.
+  * Tests runs should be consistent, and tests should pass or fail consistently *by design*. 
+    Tests should be pretty fast and cheap in terms of resources required to run.
+  * We are not currently on bazel but will likely switch to it in the future   
+
+"Ability to iterate" - critically - depends on the ability to see proposed changes in running product form. 
+  * Every successful branch build results in a deployment of the product, under a url named after the branch. 
+  * This makes changes easy to verify, review, and critique, by others.
+
+## Details
+
+* Builds execute via Github Actions
+* Note there are different builds: master vs off-master
+* Typechecking, linting, and test runs execute in the off-master build but 
+  don't block deployment (whereas in the master build, such failures will block deployment)
+* A successful master build will also replace what's served by [program.world](https://program.world/)
```

Comment:
Is the only way to rollback to revert? We should probably have some tooling around rolling back if we notice a problem... should we have some sort of lifetime each master commit is deployed so we don’t queue up a bunch of changes? I.e. master is only deployed if the previous commit has been on for 2 hours? (Weird, I know just trying to get a handle around merge yo master deploys to production with no staging... the type of bug that will be missed is if merging a branch not rebased off master. Maybe we make that a requirement?)

Reply:
> We should probably have some tooling around rolling back if we notice a problem

> I.e. master is only deployed if the previous commit has been on for 2 hours? 

To me the whole game is catching issues as close as possible in time to when the relevant code is integrated. So either you revert (if possible), or roll forward (if necessary, e.g. the commit involved a data migration, and that's now out in the wild).

If the user base becomes at all large I would want some sort of release process that separates out polls of users by stability. So you could "subscribe" to testing vs stable channels. Some system like that.

Did I address everything you had in mind here?

Reply:
yeah, basically if we succeed we will need to evolve this processes, but for now it's perfect.